docs/proposals/201911_thanos_query_mem_limits.md

```diff
@@ -0,0 +1,299 @@
+
+        // check with local limiter first
+        if err := localLimiter.Add(frameSize); err != nil {
+            return err
+        }
+
+        // now add to total
+        if err := queryTotalLimiter.Add(frameSize); err != nil {
+            return err
+        }
+    }
+}
+```
+
+## Verification
+
+Verification was done by instrumenting code and performing pprof heap dumps and comparing pprof numbers with our calculated numbers. These show as very close.
+
+Instrumentation is performed using custom code, which is now extracted into a separate package:  https://github.com/ppanyukov/go-dump  (note, it’s work in progress still).
+
+Example of instrumented figures:
+
+**// Store Gateway - Pipe Limit**
+
+```
+WRITTEN HEAP DUMP TO /Users/philip/thanos/github.com/ppanyukov/thanos-oom/heap-blockSeries-13-before.pb.gz
+MEM PROF DIFF:    	blockSeries 	blockSeries - AFTER 	-> Delta
+    AllocObjects: 	14.04K 		14.88K 			-> 863
+    AllocBytes  : 	3.86G 		4.27G 			-> 424.12M
+    InUseObjects: 	3.36K 		4.46K 			-> 1.09K
+    InUseBytes  : 	792.25M 	1.03G 			-> 265.61M
+
+LOCAL LIMITER: limit=99.95GB, current=258.01MB
+WRITTEN HEAP DUMP TO /Users/philip/thanos/github.com/ppanyukov/thanos-oom/heap-blockSeries-13-after.pb.gz
+```
+
+See that pprof reports `265.61M` and our calculations report `258.01MB`. 
+
+Running the pprof tool itself to compare the “official” snapshots, shows: `Showing nodes accounting for 220.65MB, 71.34% of 309.30MB total`, which is very close to our own reported figures.
+
+
+**// Store Gateway - Total Limit**
+
+```
+MEM PROF DIFF:    	Series 		Series - AFTER 	-> Delta
+    AllocObjects: 	12.36K 		15.70K 		-> 3.34K
+    AllocBytes  : 	3.00G 		4.52G 		-> 1.53G
+    InUseObjects: 	1.03K 		5.00K 		-> 3.97K
+    InUseBytes  : 	237.00M 	1.17G 		-> 962.28M
+
+TOTAL LIMITER: limit=99.95GB, current=919.95MB
+```
+
+Here pprof reports total merged result as `962.28M` and our calculations show `919.95MB`. Again, very close.
+
+**// Querier**
+
+```
+MEM PROF DIFF:    	Query 	Query - AFTER 	-> Delta
+    AllocObjects: 	93.59K 	116.51K 	-> 22.92K
+    AllocBytes  : 	5.64G 	7.72G 		-> 2.08G
+    InUseObjects: 	7 	23.11K 		-> 23.10K
+    InUseBytes  : 	3.96M 	916.06M 	-> 912.10M
+
+TOTAL LIMITER: limit=1.55GB, current=912.32MB
+```
+
+Similarly, pprof is `912.10M` and our calculation is `912.32MB` which is very close.
+
+Again, comparing official pprof snapshots, shows use this: `Showing nodes accounting for 800.25MB, 71.97% of 1111.96MB total`. 
+
+
+## Calculations.
+
+The calculations were done by approximations.
+
+The approach to this is using sizes of structs and objects that make up the query results. Code is like this:
+
+**Querier: receiving frames**
+
+```golang
+// Used in startStreamSeriesSet(), pkg/store/proxy.go, 
+getApproxSize := func(r *storepb.SeriesResponse) int64 {
+    size := int64(0)
+    size += int64(unsafe.Sizeof(storepb.SeriesResponse{}))
+    size += int64(len(r.XXX_unrecognized))
```

Comment:
Didn't we disable those fields?

Reply:
We did, but proposal was written *before* that PR was merged :)

Obviously, `XXX_unrecognized` won't be taken into account now.